### DISCUSSION TITLE
Usage with `vtsls` produces significant delays.

### DISCUSSION THREAD

**Discussion by KaranAhlawat** [joaotavora/eglot#1178] [Q&A] [answered]
I didn't think I should create a bug report just yet, so here I am opening a Q&A discussion.
I've wanted to try the `vtsls` language server (https://github.com/yioneko/vtsls) for TS, JS, and {J,T}SX.

I've configured the `eglot-server-programs` to use the correct command and the correct language ID for this purpose.

However, when using the server, the completions take a significant amount of time to pop up, and there are many micro-stutters while typing. It can take anywhere from 2-3 seconds for the completions to pop up (either with the built in Completions UI or Corfu).

I cannot seem to figure out why. I had previously raised an issue with the `vtsls` repository here: https://github.com/yioneko/vtsls/issues/49

As can be seen, the completions work fine and fast with Helix and NeoVim.

Any help would be appreciated :)

**Comment by joaotavora** [comment]
Karan Ahlawat ***@***.***> writes:

> I didn't think I should create a bug report just yet, so here I am
> opening a Q&A discussion.

That's fine.  That's exactly what you should do.

> As can be seen, the completions work fine and fast with Helix and NeoVim.
>
> Any help would be appreciated :)

You can share your Eglot event logs here, use M-x eglot-events buffer.

Maybe Eglot is making too many requests?  Or not making the right
request at the right time?

Else I'll need some way to reproduce this locally.

You can also, if you know how to, share a dockerfile that I can build
and which will contain a language server installation and a project
where you reproduce the problem.  I can then connect to the docker
machine with Emacs via TRAMP (I've learned to do that recently), and
hopefully reproduce the sluggishness.

João

**Reply by KaranAhlawat** [reply]
Hello! I was able to fix the slow completions with setting the `eglot-events-buffer-size` to 0. Now they function just as fast as VSCode or Neovim.

A persistent issue though, is that when using `vtsls`, either with the completion UI of emacs or through corfu, inserting of the completion candidates fails. This here is the stacktrace.

```
Debugger entered--Lisp error: (wrong-type-argument number-or-marker-p nil)
  eglot--lsp-position-to-point(nil nil)
  eglot--range-region(nil)
  #f(compiled-function (proxy status) #<bytecode 0x504426d658f0112>)("import" finished)
  completion--done("import" finished)
  #f(compiled-function (start end choice) #<bytecode -0x4131ccfc2bc9583>)(1 4 "import")
  choose-completion-string("import" #<buffer test.ts> (1 4) #f(compiled-function (start end choice) #<bytecode -0x4131ccfc2bc9583>))
  choose-completion(13 nil)
  funcall-interactively(choose-completion 13 nil)
  command-execute(choose-completion)
```

As can be seen, `eglot--range-region` is being called with a nil argument. Please guide whether this is a language server bug or an eglot bug, so I may open an issue accordingly.

**Reply by narendraj9** [reply]
I have experienced this (only the first issue and not the one with completions) as well, mainly with large projects. If there are a large number of messages being exchanged between Emacs and the language server, logging them to the `eglot-events-buffer` seems to cause significant delays. I wonder why that is the case because the communication with the language server process is already happening with a buffer that Emacs is writing to.

**Reply by yioneko** [reply]
@KaranAhlawat It's the bug from server. Fixed in vtsls 0.1.14. Please upgrade the server to see if the issue has been resolved.

**Reply by KaranAhlawat** [reply]
@yioneko Ah okay! Will check in a bit. I didn't open an issue on the vtsls repository yet because I didn't want to have the issue in the incorrect place.

**Reply by KaranAhlawat** [reply]
The update to the server now works as expected, although I still think it's a little slower than VSCode to popup the completions. But maybe that's just a limitation of tsserver being used outside of VSCode that cannot be overcome. Overall, it's working fine now!

Thank you @yioneko and @joaotavora.

**Reply by yioneko** [reply]
> I still think it's a little slower than VSCode to popup the completions.

That's what I tried to address in https://github.com/yioneko/vtsls/pull/50. You can test it early by build source code (welcome!) or wait for next release.

**Comment by KaranAhlawat** [comment, marked as answer]
I think we should move this discussion to https://github.com/yioneko/vtsls/issues/52. Would be highly appreciated if someone knowledgeable about Eglot internals could give their input.
